File: litedb-java/src/main/java/com/litedb/lsm/AtomicityDemo.java
package com.litedb.lsm;

import com.litedb.engine.WriteOp;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Comparator;
import java.util.List;

/**
 * AtomicityDemo — shows that {@link LSMEngine#writeBatch} is all-or-nothing across a crash.
 *
 * A batch is framed in the WAL between BEGIN and COMMIT markers. On recovery a batch is applied
 * only if its COMMIT is present; a batch logged without a COMMIT (a crash mid-write) is discarded
 * entirely. This is what keeps a row and its index entries from ever diverging.
 */
public final class AtomicityDemo {

    public static void main(String[] args) throws Exception {
        Path dir = Files.createTempDirectory("litedb_atomic_demo_");
        System.out.println("=== Atomic write batches (WAL BEGIN/COMMIT) ===\n");

        // Engine 1: commit one batch, then log an UNCOMMITTED batch and "crash".
        LSMEngine e1 = new LSMEngine(dir.toString());
        e1.writeBatch(List.of(WriteOp.put("a", "1"), WriteOp.put("b", "2")));
        System.out.println("committed batch:                       a=1, b=2");
        e1.writeBatchSimulateCrash(List.of(WriteOp.put("c", "3"), WriteOp.put("d", "4")));
        System.out.println("logged but NOT committed (crash sim):  c=3, d=4");
        System.out.println("...crash: no clean close, MemTable lost, WAL kept (fsync'd)\n");
        // Deliberately do NOT call e1.close().

        // Engine 2: recover from the same data dir.
        LSMEngine e2 = new LSMEngine(dir.toString());
        System.out.println("after recovery:");
        System.out.println("  GET a = " + e2.get("a") + "   (committed   -> present)");
        System.out.println("  GET b = " + e2.get("b") + "   (committed   -> present)");
        System.out.println("  GET c = " + e2.get("c") + " (uncommitted -> discarded)");
        System.out.println("  GET d = " + e2.get("d") + " (uncommitted -> discarded)");
        e2.close();

        Files.walk(dir).sorted(Comparator.reverseOrder()).map(Path::toFile).forEach(File::delete);
        System.out.println("\n[Atomicity demo complete] — the batch committed all-or-nothing.");
    }
}

File: litedb-python/atomicity_demo.py
"""atomicity_demo.py — shows LSMEngine.write_batch is all-or-nothing across a crash."""

import os
import shutil
import sys
import tempfile

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import _loader  # noqa: F401, E402
from lsm_engine import LSMEngine  # type: ignore
from storage_engine import WriteOp  # type: ignore


def main():
    d = tempfile.mkdtemp(prefix="litedb_atomic_")
    print("=== Atomic write batches (WAL BEGIN/COMMIT) ===\n")

    e1 = LSMEngine(d)
    e1.write_batch([WriteOp.put("a", "1"), WriteOp.put("b", "2")])
    print("committed batch:                      a=1, b=2")
    e1.write_batch_simulate_crash([WriteOp.put("c", "3"), WriteOp.put("d", "4")])
    print("logged but NOT committed (crash sim): c=3, d=4")
    print("...crash: no clean close\n")
    # deliberately do not close e1

    e2 = LSMEngine(d)
    print("after recovery:")
    print("  get a =", e2.get("a"), "  (committed   -> present)")
    print("  get b =", e2.get("b"), "  (committed   -> present)")
    print("  get c =", e2.get("c"), "(uncommitted -> discarded)")
    print("  get d =", e2.get("d"), "(uncommitted -> discarded)")
    e2.close()

    shutil.rmtree(d, ignore_errors=True)
    print("\n[Atomicity demo complete]")


if __name__ == "__main__":
    main()
